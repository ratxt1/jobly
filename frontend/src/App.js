import React, { Component } from 'react';
import './App.css';
import { withRouter } from 'react-router-dom';
import Routes from './Routes';
import NavBar from './NavBar';
import jwt from 'jsonwebtoken';
import JoblyApi from './JoblyApi'

/** this.state.currUser = {
 *    username, 
 *    first_name, 
 *    last_name, 
 *    email, 
 *    jobs: [ {id, title, company_handle, state}, ...], 
 *    photo_url
 * }
 */

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currUser: null,
      loading: true
    };
    this.clearCurrUser = this.clearCurrUser.bind(this);
    this.addCurrUser = this.addCurrUser.bind(this);
    this.updateCurrUser = this.updateCurrUser.bind(this);
  }

  async componentDidMount() {
    let currUser = this.state.currUser
    try {
      let token = window.localStorage.getItem('token');
      let { username } = jwt.decode(token);
      currUser = await JoblyApi.getUserInfo(username);
      this.setState({ currUser, loading: false });
    } catch(err) {
      this.setState({ loading: false });
      // this.props.history.push('/login');
    }
  }

  clearCurrUser() {
    window.localStorage.removeItem('token');
    this.setState({ currUser: null });
  }

  async addCurrUser(username) {
    try {
      let currUser = await JoblyApi.getUserInfo(username);
      this.setState({
        currUser,
      });
    } catch(err) {
    } 
  }

  updateCurrUser(updatedUser) {
    this.setState({currUser: updatedUser})
  }

  render() {
    if (this.state.loading) {
      return (
        <div className="loading">
          <i className="fas fa-4x fa-spinner fa-spin" />
        </div>
      );
    }
    return (     
      <div className="App">
          <NavBar currUser={this.state.currUser} logoutUser={this.clearCurrUser} />
          <Routes currUser={this.state.currUser} addCurrUser={this.addCurrUser} updateCurrUser={this.updateCurrUser} />
      </div>
    )
  }
}

export default withRouter(App);
// withRouter enables access to rtProps, need to wrap App in BrowserRouter in index.js
