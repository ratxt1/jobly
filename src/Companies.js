import React, { Component } from 'react';
import JoblyApi from './JoblyApi';
import CompanyCard from './CompanyCard';
import {Redirect} from "react-router-dom"
import SearchBar from './SearchBar';
// import './Companies.css';

const ERROR_STATE = {
    errorMessage: "Something went wrong!",
    loading: false
};


class Companies extends Component {
    constructor(props) {
        super(props);
        this.state = {
            companies: [],
            loading: true,
            errorMessage: ""
        }
        this.searchCompanies = this.searchCompanies.bind(this)
    }

    async componentDidMount() {
        try {
            let companies = await JoblyApi.getCompanies()
            this.setState({
                companies,
                loading: false
            })
        } catch(err) {
            this.setState(ERROR_STATE);
        }
        
    }

    async searchCompanies(searchTerm) {
        try {
            let companies = await JoblyApi.getFilteredCompanies(searchTerm)
            this.setState({ companies })
        } catch(err) {
            this.setState(ERROR_STATE);
        }
        
    }

    render() {
        if(this.props.currUser === null){
            return <Redirect to="/login"/>
        }
        const companies = this.state.companies.map(c => <CompanyCard key={c.handle} {...c} />);

        if (this.state.loading) {
            return (
                <div className="loading">
                    <i className="fas fa-4x fa-spinner fa-spin" />
                </div>
            );
        }

        if (this.state.errorMessage) {
            return (
                <p>{this.state.errorMessage}</p>
            )
        }

        return (
            <div className="Companies row">
                <div className="col-8 offset-2">
                    <SearchBar search={this.searchCompanies} />
                    {companies}
                </div>
            </div>
        )
    }
}

export default Companies;
