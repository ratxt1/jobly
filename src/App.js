import React, { Component } from 'react';
import './App.css';
import { BrowserRouter, withRouter } from 'react-router-dom';
import Routes from './Routes';
import NavBar from './NavBar';
import jwt from 'jsonwebtoken';
import JoblyApi from './JoblyApi'

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currUser: null,
      loading: true
    };
    this.clearCurrUser = this.clearCurrUser.bind(this);
    this.addCurrUser = this.addCurrUser.bind(this);
    this.updateCurrUser = this.updateCurrUser.bind(this);
  }

  async componentDidMount() {
    let currUser = this.state.currUser
    try {
      let token = window.localStorage.getItem('token');
      let { username } = jwt.decode(token);
      currUser = await JoblyApi.getUserInfo(username);
      // currUser = {username, first_name, last_name, email, jobs[{id, title, company_handle, state}], photo_url}
      this.setState({
        currUser,
        loading: false
      });
    } catch(err) {
      this.setState({
        loading: false
      });
      // this.props.history.push('/login');
    }
  }

  clearCurrUser() {
    window.localStorage.removeItem('token');
    this.setState({ currUser: null });
  }

  async addCurrUser(username) {
    console.log("Add curruser username: ", username)
    try {
      let currUser = await JoblyApi.getUserInfo(username);
      this.setState({
        currUser,
      });
    } catch(err) {
    } 
  }

  updateCurrUser(updatedUser) {
    this.setState({currUser: updatedUser})
  }

  render() {
    if (this.state.loading) {
      return (
        <div className="loading">
          <i className="fas fa-4x fa-spinner fa-spin" />
        </div>
      );
    }
    return (     
      <div className="App">
          <NavBar currUser={this.state.currUser} logoutUser={this.clearCurrUser} />
          <Routes currUser={this.state.currUser} addCurrUser={this.addCurrUser} updateCurrUser={this.updateCurrUser} />
      </div>
    )
  }
}

export default withRouter(App);

/** componentDidMount
 * localStorage.getItem('token')
 *
 * JoblyAPI has
 *
 * token payload has username
 * API call to get /users/username (authenticates, returns users info)
 *
 * LATER - returned info should include the jerbs user has applied to
 */