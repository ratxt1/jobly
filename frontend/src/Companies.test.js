/** TESTING PLAN!
 * 
 * Questions/concerns 
 * - confirm that we should mock API calls 
 * 
 * App
 * - Renders nav bar
 * - LATER: renders correct UI based on if youre logged in vs. not 
 * 
 * Companies
 * - componentDidMount ran correctly (changes state)
 * - renders all companies (snapshot?)
 * - clicking a company shows companies detail 
 *   - check that CompanyCard component has been rendered 
 *   (check wrapper.state.companyData = that comapnys info obj
 * 
 * Company
 *  - componentDidMount ran correctly (setsState with requested companies data)
 * 
 * CompanyCard
 * - shallow/snapshot
 * - uses default image if company does not have an image
 * 
 * JobCard
 *  - shallow/snapshot
 * 
 * SearchBar
 * 
 */

import React from "react";
import { shallow, mount } from "enzyme";
import rrd, { MemoryRouter, BrowserRouter } from 'react-router-dom';
import toJson from "enzyme-to-json";
import Companies from "./Companies";
import App from './App'

let wrapper;


it("renders without crashing", function() {
    wrapper = mount(
        <MemoryRouter>
          <Companies/>
        </MemoryRouter>
    );
});
  
// snapshot test
it("matches snapshot", function() {
    wrapper = mount(
        <MemoryRouter>
          <Companies/>
        </MemoryRouter>
    );
    let companies = wrapper.find(Companies);
    let serialized = toJson(companies);
    expect(serialized).toMatchSnapshot();
});

it("runs componentDidMount", function(){
    const mockComponentDidMount = jest.spyOn(
        Companies.prototype,
        "componentDidMount"
    )
    wrapper = mount(
        <MemoryRouter>
          <Companies/>
        </MemoryRouter>
    );
    expect(mockComponentDidMount).toHaveBeenCalled();
})

it("runs componentDidMount with updated state", async function(){
    // NOTES: in App we've defined BrowserRouter to have routes in it. 
    // When we're testing we need to overwrite what the default BrowserRouter does.
    // In the example below we are returning the App (thats the child)

    // initialEntries brings you to that "route", 
    // if you're mounting App you can use initialEntries to bring you somewhere

    wrapper = mount(
        <MemoryRouter initialEntries={[ '/companies' ]}>
          <Companies/>
        </MemoryRouter>
      );
    
    let companies = wrapper.find(Companies).instance();

    expect(companies.state).toEqual({ companiesData: [], loading: true });

    await companies.componentDidMount();

    expect(companies.state.loading).toEqual(false);
})

