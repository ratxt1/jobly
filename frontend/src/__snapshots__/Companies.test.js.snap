// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Companies>
  <div
    className="loading"
  >
    <i
      className="fas fa-4x fa-spinner fa-spin"
    />
  </div>
</Companies>
`;
